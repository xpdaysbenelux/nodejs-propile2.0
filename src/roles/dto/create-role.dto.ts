import { IsString, ValidateNested, IsNotEmpty } from 'class-validator';
import { Type } from 'class-transformer';

import { PermissionsDto } from './permissions.dto';

export class CreateRoleRequest {
    @IsNotEmpty()
    @IsString()
    readonly name: string;

    @Type(() => PermissionsDto)
    @ValidateNested()
    readonly permissions: PermissionsDto;
}
