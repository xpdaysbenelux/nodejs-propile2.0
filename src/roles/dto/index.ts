export * from './create-role.dto';
export * from './get-role.dto';
export * from './get-roles.dto';
export * from './permissions.dto';
export * from './update-role.dto';
