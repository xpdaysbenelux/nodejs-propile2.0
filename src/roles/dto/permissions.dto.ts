import { IsBoolean, IsOptional } from 'class-validator';

class RolePermissionsDto {
    @IsBoolean()
    @IsOptional()
    readonly view?: boolean;

    @IsBoolean()
    @IsOptional()
    readonly edit?: boolean;
}

class UserPermissionsDto {
    @IsBoolean()
    @IsOptional()
    readonly view?: boolean;

    @IsBoolean()
    @IsOptional()
    readonly edit?: boolean;
}

export class PermissionsDto {
    @IsOptional()
    readonly roles?: RolePermissionsDto;

    @IsOptional()
    readonly users?: UserPermissionsDto;
}
