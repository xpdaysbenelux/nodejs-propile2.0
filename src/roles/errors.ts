import {
    BadRequestException,
    NotFoundException,
    MethodNotAllowedException,
} from '@nestjs/common';

export class RoleNameAlreadyInUse extends BadRequestException {
    constructor() {
        super(
            'A role with this name already exists',
            'ROLE_NAME_ALREADY_IN_USE',
        );
    }
}

export class RoleNotFound extends NotFoundException {
    constructor() {
        super('The given role was not found', 'ROLE_NOT_FOUND');
    }
}

export class RoleInUse extends MethodNotAllowedException {
    constructor() {
        super(
            'It is not possible to delete a role that is being used',
            'ROLE_IN_USE',
        );
    }
}
