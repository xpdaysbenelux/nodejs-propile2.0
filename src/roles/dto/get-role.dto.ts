import { Type } from 'class-transformer';

import { BaseEntityResponse } from '../../_shared/dto';
import { PermissionsDto } from './permissions.dto';

export class RoleResponse extends BaseEntityResponse {
    readonly name: string;
    @Type(() => PermissionsDto)
    readonly permissions: PermissionsDto;
}
