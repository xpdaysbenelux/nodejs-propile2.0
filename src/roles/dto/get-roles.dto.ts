import { IsOptional, IsEnum } from 'class-validator';

import { PagingQuery, PagingMeta } from '../../_shared/dto';
import { RoleResponse } from './get-role.dto';

export enum RolesSortColumns {
    Name = 'name',
    CreatedAt = 'createdAt',
    UpdatedAt = 'updatedAt',
    CreatedBy = 'createdBy',
    UpdatedBy = 'updatedBy',
}

export class GetRolesRequestQuery extends PagingQuery {
    @IsOptional()
    @IsEnum(RolesSortColumns)
    readonly sortBy?: RolesSortColumns;
}

export class GetRolesResponse {
    readonly meta: PagingMeta;
    readonly data: RoleResponse[];
}
