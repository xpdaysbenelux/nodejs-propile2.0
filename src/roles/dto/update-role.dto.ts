import { IsString, ValidateNested, IsOptional, IsUUID } from 'class-validator';
import { Type } from 'class-transformer';
import { ApiProperty } from '@nestjs/swagger';

import { PermissionsDto } from './permissions.dto';

export class UpdateRoleRequest {
    @IsOptional()
    @IsString()
    readonly name?: string;

    @IsOptional()
    @Type(() => PermissionsDto)
    @ValidateNested()
    readonly permissions?: PermissionsDto;
}

export class RoleIdParam {
    @ApiProperty({ required: true })
    @IsUUID('4')
    readonly roleId: string;
}
