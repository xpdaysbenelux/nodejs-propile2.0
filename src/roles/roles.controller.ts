import {
    Controller,
    Post,
    Body,
    Get,
    Query,
    UseGuards,
    Patch,
    Param,
    Delete,
    HttpCode,
    HttpStatus,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import {
    CreateRoleRequest,
    GetRolesResponse,
    GetRolesRequestQuery,
    RoleResponse,
    RoleIdParam,
    UpdateRoleRequest,
} from './dto';
import { RolesService } from './roles.service';
import { RolesQueries } from './roles.queries';
import {
    AuthenticatedGuard,
    RequiredPermissionsGuard,
} from '../_shared/guards';
import { UserSession, RequiredPermissions } from '../_shared/decorators';
import { IUserSession } from '../_shared/constants';

@UseGuards(AuthenticatedGuard)
@ApiTags('roles')
@Controller('roles')
export class RolesController {
    constructor(
        private readonly rolesService: RolesService,
        private readonly rolesQueries: RolesQueries,
    ) {}

    @RequiredPermissions({ roles: { view: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get()
    getRoles(@Query() query: GetRolesRequestQuery): Promise<GetRolesResponse> {
        return this.rolesQueries.getRoles(query);
    }

    @RequiredPermissions({ roles: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Post()
    async createRole(
        @Body() body: CreateRoleRequest,
        @UserSession() session: IUserSession,
    ): Promise<RoleResponse> {
        const roleId = await this.rolesService.createRole(body, session);
        return this.rolesQueries.getRole(roleId);
    }

    @RequiredPermissions({ roles: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Patch(':roleId')
    async updateRole(
        @Body() body: UpdateRoleRequest,
        @Param() params: RoleIdParam,
        @UserSession() session: IUserSession,
    ): Promise<RoleResponse> {
        // Only the provided properties will be updated
        const roleId = await this.rolesService.updateRole(
            body,
            params.roleId,
            session,
        );
        return this.rolesQueries.getRole(roleId);
    }

    @RequiredPermissions({ roles: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @HttpCode(HttpStatus.NO_CONTENT)
    @Delete(':roleId')
    deleteRole(@Param() params: RoleIdParam): Promise<void> {
        return this.rolesService.deleteRole(params.roleId);
    }
}
