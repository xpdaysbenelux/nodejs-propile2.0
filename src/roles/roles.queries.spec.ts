import { Test } from '@nestjs/testing';
import * as faker from 'faker';

import { RolesQueries } from './roles.queries';
import { DatabaseModule, RoleRepository } from '../database';
import { GetRolesRequestQuery, RolesSortColumns } from './dto';
import { SortDirection } from '../_shared/constants';

describe('RolesQueries', () => {
    let roleRepository: RoleRepository;
    let rolesQueries: RolesQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [RolesQueries],
        }).compile();

        rolesQueries = module.get(RolesQueries);
        roleRepository = module.get(RoleRepository);
    });

    afterAll(() => {
        roleRepository.manager.connection.close();
    });

    describe('getRole', () => {
        it('should return the requested role correctly', async () => {
            const result = await rolesQueries.getRole(
                '0c1510ab-5dca-41e2-8912-ee165140ae90',
            );
            expect(result).toMatchSnapshot();
        });

        it('should return nothing if the requested role does not exist', async () => {
            const result = await rolesQueries.getRole(faker.random.uuid());
            expect(result).toMatchSnapshot();
        });
    });

    describe('getRoles', () => {
        const testQueries: GetRolesRequestQuery[] = [
            {},
            { skip: 2, take: 2 },
            { search: 'admin' },
            { search: 'SUPPORT' },
            {
                sortBy: RolesSortColumns.CreatedAt,
                sortDirection: SortDirection.Ascending,
            },
        ];

        testQueries.forEach((query, index) => {
            it(`should return a paged list of roles: #${index}`, async () => {
                const result = await rolesQueries.getRoles(query);
                expect(result).toMatchSnapshot();
            });
        });

        it('should return an empty list if no roles found', async () => {
            const result = await rolesQueries.getRoles({
                search: 'nonsensicaljibberish',
            });
            expect(result).toMatchSnapshot();
        });
    });
});
