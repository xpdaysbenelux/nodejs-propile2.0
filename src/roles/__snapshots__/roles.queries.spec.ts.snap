// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesQueries getRole should return nothing if the requested role does not exist 1`] = `undefined`;

exports[`RolesQueries getRole should return the requested role correctly 1`] = `
Role {
  "createdAt": 2019-06-20T08:12:59.730Z,
  "createdBy": null,
  "id": "0c1510ab-5dca-41e2-8912-ee165140ae90",
  "name": "Admin",
  "permissions": Object {
    "roles": Object {
      "edit": true,
      "view": true,
    },
    "users": Object {
      "edit": true,
      "view": true,
    },
  },
  "updatedAt": 2019-06-20T08:12:59.730Z,
  "updatedBy": null,
}
`;

exports[`RolesQueries getRoles should return a paged list of roles: #0 1`] = `
Object {
  "data": Array [
    Role {
      "createdAt": 2019-10-20T08:12:59.730Z,
      "createdBy": "development@icapps.com",
      "id": "d02a468f-0225-463d-a9a4-8c7d05cbe62c",
      "name": "Support",
      "permissions": Object {
        "roles": Object {
          "edit": false,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-12-20T08:12:59.730Z,
      "updatedBy": "development@icapps.com",
    },
    Role {
      "createdAt": 2019-08-20T08:12:59.730Z,
      "createdBy": "development@icapps.com",
      "id": "81388934-7613-4f66-a225-a0d246f65e61",
      "name": "Medewerker",
      "permissions": Object {
        "roles": Object {
          "edit": false,
          "view": false,
        },
        "users": Object {
          "edit": false,
          "view": true,
        },
      },
      "updatedAt": 2019-09-20T08:12:59.730Z,
      "updatedBy": "development@icapps.com",
    },
    Role {
      "createdAt": 2019-06-20T08:12:59.730Z,
      "createdBy": null,
      "id": "0c1510ab-5dca-41e2-8912-ee165140ae90",
      "name": "Admin",
      "permissions": Object {
        "roles": Object {
          "edit": true,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-06-20T08:12:59.730Z,
      "updatedBy": null,
    },
  ],
  "meta": Object {
    "count": 3,
    "skip": 0,
    "totalCount": 3,
  },
}
`;

exports[`RolesQueries getRoles should return a paged list of roles: #1 1`] = `
Object {
  "data": Array [
    Role {
      "createdAt": 2019-06-20T08:12:59.730Z,
      "createdBy": null,
      "id": "0c1510ab-5dca-41e2-8912-ee165140ae90",
      "name": "Admin",
      "permissions": Object {
        "roles": Object {
          "edit": true,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-06-20T08:12:59.730Z,
      "updatedBy": null,
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 2,
    "totalCount": 3,
  },
}
`;

exports[`RolesQueries getRoles should return a paged list of roles: #2 1`] = `
Object {
  "data": Array [
    Role {
      "createdAt": 2019-06-20T08:12:59.730Z,
      "createdBy": null,
      "id": "0c1510ab-5dca-41e2-8912-ee165140ae90",
      "name": "Admin",
      "permissions": Object {
        "roles": Object {
          "edit": true,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-06-20T08:12:59.730Z,
      "updatedBy": null,
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`RolesQueries getRoles should return a paged list of roles: #3 1`] = `
Object {
  "data": Array [
    Role {
      "createdAt": 2019-10-20T08:12:59.730Z,
      "createdBy": "development@icapps.com",
      "id": "d02a468f-0225-463d-a9a4-8c7d05cbe62c",
      "name": "Support",
      "permissions": Object {
        "roles": Object {
          "edit": false,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-12-20T08:12:59.730Z,
      "updatedBy": "development@icapps.com",
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`RolesQueries getRoles should return a paged list of roles: #4 1`] = `
Object {
  "data": Array [
    Role {
      "createdAt": 2019-06-20T08:12:59.730Z,
      "createdBy": null,
      "id": "0c1510ab-5dca-41e2-8912-ee165140ae90",
      "name": "Admin",
      "permissions": Object {
        "roles": Object {
          "edit": true,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-06-20T08:12:59.730Z,
      "updatedBy": null,
    },
    Role {
      "createdAt": 2019-08-20T08:12:59.730Z,
      "createdBy": "development@icapps.com",
      "id": "81388934-7613-4f66-a225-a0d246f65e61",
      "name": "Medewerker",
      "permissions": Object {
        "roles": Object {
          "edit": false,
          "view": false,
        },
        "users": Object {
          "edit": false,
          "view": true,
        },
      },
      "updatedAt": 2019-09-20T08:12:59.730Z,
      "updatedBy": "development@icapps.com",
    },
    Role {
      "createdAt": 2019-10-20T08:12:59.730Z,
      "createdBy": "development@icapps.com",
      "id": "d02a468f-0225-463d-a9a4-8c7d05cbe62c",
      "name": "Support",
      "permissions": Object {
        "roles": Object {
          "edit": false,
          "view": true,
        },
        "users": Object {
          "edit": true,
          "view": true,
        },
      },
      "updatedAt": 2019-12-20T08:12:59.730Z,
      "updatedBy": "development@icapps.com",
    },
  ],
  "meta": Object {
    "count": 3,
    "skip": 0,
    "totalCount": 3,
  },
}
`;

exports[`RolesQueries getRoles should return an empty list if no roles found 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;
